device/devicegate/filter.go

```diff
@@ -0,0 +1,263 @@
+	claimsLocation      = "claims"
+)
+
+var (
+	emptyVal = struct{}{}
+)
+
+type FilterGateOption func(*FilterGate)
+
+type FilterGate struct {
+	filters        map[string]map[interface{}]struct{}
+	allowedFilters map[string]struct{}
+	lock           sync.RWMutex
+}
+
+type filterRequest struct {
+	Key    string
+	Values []interface{}
+}
+
+func NewFilterGate(options ...FilterGateOption) *FilterGate {
+	fg := &FilterGate{
+		filters:        make(map[string]map[interface{}]struct{}),
+		allowedFilters: make(map[string]struct{}),
+	}
+
+	for _, o := range options {
+		o(fg)
+	}
+
+	return fg
+
+}
+
+func WithAllowedFilters(allowedFilters map[string]struct{}) FilterGateOption {
+	return func(fg *FilterGate) {
+		if allowedFilters != nil {
+			fg.allowedFilters = allowedFilters
+		} else {
+			fg.allowedFilters = make(map[string]struct{})
+		}
+	}
+}
+
+func (f *FilterGate) PrettifyFilters() map[string][]interface{} {
+	copy := make(map[string][]interface{})
+
+	f.lock.RLock()
+	for k, v := range f.filters {
+		var filters []interface{}
+		for filter := range v {
+			filters = append(filters, filter)
+		}
+
+		copy[k] = filters
+	}
+	f.lock.RUnlock()
+
+	return copy
+}
+
+func (f *FilterGate) FiltersCopy() map[string]map[interface{}]struct{} {
+	filtersCopy := make(map[string]map[interface{}]struct{})
+
+	f.lock.RLock()
+	for key, val := range f.filters {
+		valCopy := make(map[interface{}]struct{})
+
+		for k, v := range val {
+			valCopy[k] = v
+		}
+
+		filtersCopy[key] = valCopy
+	}
+	f.lock.RUnlock()
+
+	return filtersCopy
+}
+
+func (f *FilterGate) Filters() map[string]map[interface{}]struct{} {
+	return f.filters
+}
+
+//allows for adding and removing filters
+//Will completely overwrite current filters in place
+func (f *FilterGate) EditFilters(key string, values []interface{}, add bool) bool {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	if add {
+		valuesMap := make(map[interface{}]struct{})
+
+		for _, v := range values {
+			valuesMap[v] = emptyVal
+		}
+
+		f.filters[key] = valuesMap
+
+		return true
+	} else {
+
+		_, ok := f.filters[key]
+
+		if ok {
+			delete(f.filters, key)
+			return true
+		}
+
+		return false
+	}
+}
+
+func (f *FilterGate) AllowConnection(d device.Interface) (bool, device.MatchResult) {
+	f.lock.RLock()
+	defer f.lock.RUnlock()
+
+	for filterKey, filterValues := range f.filters {
+
+		// check if filter is in claims
+		if claimsMatch(filterKey, filterValues, d.Metadata()) {
+			return false, device.MatchResult{
+				Location: metadataMapLocation,
+				Key:      filterKey,
+			}
+		}
+
+		// check if filter is in metadata map
+		if metadataMapMatch(filterKey, filterValues, d.Metadata()) {
+			return false, device.MatchResult{
+				Location: claimsLocation,
+				Key:      filterKey,
+			}
+		}
+
+	}
+
+	return true, device.MatchResult{}
+}
+
+func metadataMapMatch(keyToCheck string, filterValues map[interface{}]struct{}, m *device.Metadata) bool {
+	metadataVal := m.Load(keyToCheck)
```

Comment:
If you make a custom type for the filters, something like `type Filters map[string]map[interface{}]struct{}`, then functions like this can simply be methods of that type.